Second level model

MartinPerez opened the pull request:
Here is the second level model with one simple example (second level analysis only) on the mixed gambles dataset (which is more a sanity check than an actual example due to the lack of a nice contrast to picture the results). I will create a new PR for a complete first and second level example on a bids dataset. The second level model object is quite flexible at the moment, it accepts list of first level models, the pandas dataframe discussed with @bthirion  and a simple list of images, then the other arguments must be matched accordingly. I hope the documentation is clear in this respect.

At the moment the second level only includes a simple OLS regression that only takes into account effect maps (RFX). Future PRs would be done for more complex models.

I included the patsy formulas functionality for the contrasts, it was a very small adaptation of the code. So we already have formulas for contrasts in both first and second level objects.

Is built on top of the first level model, so it should be compared with #72 instead of master until #72 is merged.

All tests included too.

salma1601 commented:
Hi @MartinPerez 
I guess this PR is a work in progress. Is it planned to be completed soon ?
I need second level pipeline and I am hesitating between nistats and spm

MartinPerez commented:
Hi @salma1601 actually the PR is ready and functional, in testing and review phase. With small typos like the one you noticed in the example due to the radical API changes. I am currently testing it with my own datasets to also notice if some additional API changes would be useful.

Being responsible I think if you need to do the analysis right now I would recommend to still use spm, in a couple of weeks this will likely be merged but even then it would be good to compare spm and nistats output. For example an issue was raised recently #71 showing that parametric modulators seem to be giving different results in both frameworks even though the rest is fine.

Please feel welcome to also jump in the discussion of the API changes. It would also be great if more people give a hand comparing packages with their datasets and user cases to make nistats more robust.

MartinPerez commented:
API discussion in #65

salma1601 commented:
Thanks for the quick reply.
>It would also be great if more people give a hand comparing packages with their datasets and user cases to make nistats more robust.

OK I am going to launch first level (and second level as soon as merged) on my data and warn if problems

MartinPerez commented:
in first level model compute contrast added the option to get the fixed effect contrast by default or the list of corresponding run contrasts. This will make it easy to collect per run maps for machine learning or other purposes.

MartinPerez commented:
it seems there is an issue with pandas 0.12.0 with python 2.6 in the second level design matrix creation. checking it out...

bthirion commented:
On 19/06/2016 01:04, Martin Perez-Guevara wrote:
>
> in first level model compute contrast added the option to get the 
> fixed effect contrast by default or the list of corresponding run 
> contrasts. This will make it easy to collect per run maps for machine 
> learning or other purposes.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/nistats/nistats/pull/78#issuecomment-226969791>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/AAOT1r7BSKnYH-W-MmEjWHk04ORJ1B1nks5qNHmBgaJpZM4Iz6-V>.
>
Could this be done in a separate PR ? It is very hard to ensure the 
consistency of the API and the abstractions used by this code, hence we 
need to separate the different problems / fixes / improvements.

MartinPerez commented:
ok i will separate the fixed effect feature in a different PR

MartinPerez commented:
@bthirion in the process of fixing the error on pandas version 0.12 (turns out enlargement with .loc was not supported by until 0.13) I started to test the minimum requirements and it turns out nilearn 0.1 create many import issues in the current code. Moreover nilearn 0.2 is checked for conflicts in conda such that it is not compatible with python 2.6.

In install.sh nilearn is installed directly with pip, but then this is not compatible with the minimum requirements we are announcing for nistats.

A simple annoying example is that the check_niimg api changed radically from v 0.1

I admit so far I was developing with the latest libraries installed and overlooking minimum requirements. I checked we need to do patsy 0.2.0 (support python 2.6) or then jump to 0.4.1 (first version to not announce conflicts with python 2.7 in conda)

What should we do about this? I would vote for following a latests nilearn version than 0.1 minimum requirements and going for pandas 0.13 as the minimum.

bthirion commented:
Yes, given the very early state of Nistats, it seems reasonable to support only recent versions of the dependencies. We don't want to deal with the burden of bw compatibility for pandas, nilearn etc.

MartinPerez commented:
ok, shall we also drop official support for python 2.6? many libraries do not seem to support it anymore in the conda dependencies check.

bthirion commented:
I would accept the same dependencies as Nilearn, i.e. Python 2.6 --for 
the moment.

MartinPerez commented:
travis needs to be restarted, some of the builds failed for connection problems or other unrelated travis errors.

bthirion commented on continuous_integration/show-python-packages-versions.py:
```diff
@@ -1,6 +1,7 @@
-DEPENDENCIES = ['numpy', 'scipy', 'sklearn', 'matplotlib', 'nibabel', 'pandas']
+DEPENDENCIES = ['numpy', 'scipy', 'nilearn', 'nibabel', 'pandas', 'patsy',
+                'sklearn', 'matplotlib']
 
```
Could all this be in a separate PR ?

bthirion commented on examples/plot_mixed_gambles.py:
```diff
@@ -0,0 +1,80 @@
+
+
+# Second level model #########################################
+# create first_level_input
```
this comment is a bit misleading. "create input for second-level analysis" sounds better.

bthirion commented on nistats/design_matrix.py:
```diff
@@ -443,3 +443,31 @@ def plot_design_matrix(design_matrix, rescale=True, ax=None):
+    """Infers a second level design from a maps table.
+
+    The maps table is a pandas DataFrame with at least columns 'map_name' and
+    'subject_id', confounds is also a pandas DataFrame with at least two
```
In some cases (only one map per subject), you don't want to have the subject specific regressors. I would not impose it.
Also, beware that 'second-level' could be some kind of multi-acquisition/multi-session setting.

bthirion commented on examples/plot_mixed_gambles.py:
```diff
@@ -0,0 +1,80 @@
+df_columns = [subjects_id, behavioral_target, zmap_paths]
+df_columns = zip(*df_columns)
+df_column_names = ['subject_id', 'map_name', 'effects_map_path']
+first_level_df = pd.DataFrame(df_columns, columns=df_column_names)
```
I think that we should not call it `first_level_df`: this object actually represent the second-level inputs. Actually, looking at it a bit more, I find it confusing.
I think that  `map_name` is actually more something like `experimental_condition`.

bthirion commented on nistats/design_matrix.py:
```diff
@@ -443,3 +443,31 @@ def plot_design_matrix(design_matrix, rescale=True, ax=None):
+    The maps table is a pandas DataFrame with at least columns 'map_name' and
+    'subject_id', confounds is also a pandas DataFrame with at least two
+    columns 'subject_id' and one confounder.
+    """
```
Also the `confounder` deserves clarification: conceptually, what is the conceptual difference between the confounder and the `maps_table` ?

bthirion commented on nistats/first_level_model.py:
```diff
@@ -406,7 +413,7 @@ def fit(self, run_imgs, paradigms=None, confounds=None,
                 frame_times = np.linspace(start_time, end_time, n_scans)
-                design = make_design_matrix(frame_times, paradigms[run_idx],
+                design = make_design_matrix(frame_times, events[run_idx],
                                             self.hrf_model, self.drift_model,
```
I think that I agree with the change from 'paradigms' to 'events', but please try and keep this type of changes for other PRs, since this has exactly no relationship with second-level models.

bthirion commented on nistats/first_level_model.py:
```diff
@@ -441,22 +448,27 @@ def fit(self, run_imgs, paradigms=None, confounds=None,
+            the fitted model combined with operators /*+- and numbers.
+            Nonetheless all contrasts assume the baseline of the tested effect
+            as 0, so specifying a different baseline in a formula will have no
+            effect. Please checks the patsy documentation for formula examples:
```
I am not sure that the baseline question should be discussed here (looks like a corner case).

bthirion commented on nistats/first_level_model.py:
```diff
@@ -467,23 +479,33 @@ def compute_contrast(self, contrast_def, contrast_name=None,
-        output_image : Nifti1Image
-            The desired output image
+        output : Nifti1Image
+            The desired output image from a fixed effects computation.
```
Why do emphasize the notion of fixed effect ? This seems a bit unnatural here.

bthirion commented on nistats/first_level_model.py:
```diff
@@ -467,23 +479,33 @@ def compute_contrast(self, contrast_def, contrast_name=None,
+        # We reinterpret formulas as vectors only if design has column names.
+        des_names = np.unique(self.design_matrices_[0].columns.tolist())
+        if len(des_names) == len(self.design_matrices_[0].columns):
+            di = DesignInfo(self.design_matrices_[0].columns.tolist())
```
`design_info` instead of `di`

bthirion commented on nistats/first_level_model.py:
```diff
@@ -467,23 +479,33 @@ def compute_contrast(self, contrast_def, contrast_name=None,
+                if not isinstance(con, np.ndarray):
+                    raise ValueError('can not interpret formula when design'
+                                     ' do not contain columns with unique '
+                                     'names')
```
'Cannot interpret formula when design does not contain columns with unique identifiers'

bthirion replied:
Hm, thinking a bit further: we certainly don't want design to contain columns with redundant identifiers, so this should be enforced in the design matrix construction, not here (I find this code is becoming too convoluted because we are taking care of such issues that should be handled elsewhere)

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+from .design_matrix import _create_second_level_design
+
+
+def run_glm(Y, X, n_jobs=1, verbose=0):
```
We already have a run_glm in first-level_glm. As these two guys are doing the same thing, this one should be replaced by a call to the other one.

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+    def fit(self, first_level_input, first_level_conditions=None,
+            confounds=None, design_matrix=None,
+            first_level_conditions_name=None):
+        """ Fit the GLM
```
Fit the second-level GLM

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+
+        Parameters
+        ----------
+        first_level_input: list of `FirstLevelModel` objects or pandas
```
as discussed in the example, I am afraid that this terminology will confuse users, as it can be interpreted "input to first level analysis". `second_level_input` makes more sense to me. Do you agree ?

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+        first_level_conditions: list of (str or array) or None
+            If first_level_input is a list of `FirstLevelModel` objects then
+            it is mandatory to provide a list of contrast definitions to pass
+            to the compute_contrast method of `FirstLevelModel` which can be
```
avoid 'which can be', because it is not obvious what 'which' refers to in such nested sentences.

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+
+            If first_level_input is a pandas DataFrame then a list of strings
+            is expected with the names of maps that will be part of the model.
+            None indicates that all maps should be included.
```
If None, all maps are included

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+            If first_level_input is a list of Niimg-like objects then this
+            argument will be ignored.
+
+        confounds: pandas DataFrame, optional
```
I don't like this 'confounds' terminology, as this may well be an effect of interest (e.g. the age of the subject, see may point above). Use 'regressors' instead of 'confounds'.

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+            design_matrix is provided then this argument will be ignored.
+            The resulting second level design matrix will use the same column
+            names as in the given DataFrame for confounds. At least two columns
+            are expected, "subject_id" and at least one confounder.
```
Nit-picking: avoid using future 'will' for these descriptions. This makes the formulation heavier for no benefit.

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+        design_matrix: pandas DataFrame, optional
+            Design matrix that will be used to fit the GLM. The number of rows
+            in the design matrix must agree with the size of the list in
+            first_level_conditions. Careful that the order of maps given
```
Be careful that

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+            first_level_conditions. Careful that the order of maps given
+            by first_level_conditions or inferred directly from
+            first_level_input when first_level_conditions is None correctly
+            match the order of the rows in the design matrix.
```
Ensure that the order of maps given by first_level_conditions or inferred directly from first_level_input matches the order of the rows in the design matrix.

bthirion replied:
Maybe we should think of mechanisms for such checks as this is not straightforward.

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+        first_level_conditions_name: list of str or None,
+            This argument only applies when first_level_input is a list of
+            `FirstLevelModel` objects. It specifies the name of the columns
+            corresponding to the provided first_level_conditions.
```
I find this a bit convoluted: this hints at a design problem. Esentially first_level_conditions_name and first_level_conditions point to the same object, this probably means that this one should be e.g. a dictionary

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+        """
+        # Check parameters
+        # check first level input
+        if isinstance(first_level_input, FirstLevelModel):
```
This check is a bit weird. Why not checking that this is indeed a list of FirstleveModels, image or df ?

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+            # Check FirstLevelModel objects case
+            if isinstance(first_level_input[0], FirstLevelModel):
+                if first_level_conditions is None:
+                    raise ValueError('First level models input require'
```
requires

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+                                             ' to match rows appropriately'
+                                             '. model at idx %d do not provide'
+                                             ' it. It can be set by doing '
+                                             'model.subject_id = "01"')
```
first_level_model.subject_id = "01".

nitpicking: could you make longer lines by starting a new line after the opening parenthesis ?

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+                                         ' first_level_input' % condition)
+                in_cond = first_level_input.apply(lambda x: x['map_name'] in
+                                                  first_level_conditions,
+                                                  axis=1)
```
Please avoid this: this makes the code uselessly hard to undertsand.

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+                design = _create_second_level_design(first_level_input,
+                                                     confounds)
+            # get effect maps
+            effects_maps = first_level_input['effects_map_path'].tolist()
```
Maybe this is not the right place for this comment, yet we should remember that for mixed-effects models, you typically want to have bit the first-level effects *and variance*.

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+            effects_maps = []
+            for model in first_level_input:
+                for con_idx, con_def in enumerate(first_level_conditions):
+                    con_name = 'con_%02d' % con_idx
```
should be first_level_condition_names rather than a rather arbitrary integer.

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+            for model in first_level_input:
+                for con_idx, con_def in enumerate(first_level_conditions):
+                    con_name = 'con_%02d' % con_idx
+                    if first_level_conditions_name:
```
All right. I'd rather introduce a deafult for first_level conditions_name, but to me the right aprpoach is rather to use first_level_conditions.keys()

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+                    effects_maps.append(eff_map)
+            # Get the design matrix
+            if design_matrix is None:
+                design = _create_second_level_design(design_df, confounds)
```
I find it hard to undertsand the distinction between design_df and design_matrix.
* I think that the design of _create_second_levle _design is a key issue. This function should be part of the public API, and there has to be an example dedicated only to presenting what it does.
* Renaming will be important. `design` alone is not clear enough; `design_matrix` is certainly better. `design_df` is not great; may be `design_input`

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+        if design_matrix is None:
+            self.design_matrix_ = design
+        else:
+            self.design_matrix_ = design = design_matrix
```
:-1:

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+        # We save memory if inspecting model details is not necessary
+        if self.minimize_memory:
+            for key in results:
+                results[key] = SimpleRegressionResults(results[key])
```
Let's keep this for the moment, but in the future this will have to go.

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+                         output_type='z_score'):
+        """Generate different outputs corresponding to
+        the contrasts provided e.g. z_map, t_map, effects and variance.
+        In multi-session case, outputs the fixed effects map.
```
this does not make sense here: no multi-session here (?)
Please update the whole docstring accordingly.

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+            if np.all(con_val == 0):
+                raise ValueError('Contrast is null')
+        else:
+            des_names = np.unique(self.design_matrix_.columns.tolist())
```
design_ames

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+        else:
+            des_names = np.unique(self.design_matrix_.columns.tolist())
+            if len(des_names) == len(self.design_matrix_.columns):
+                di = DesignInfo(self.design_matrix_.columns.tolist())
```
design_info

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+                di = DesignInfo(self.design_matrix_.columns.tolist())
+                con_val = di.linear_constraint(contrast_def).coefs
+            else:
+                raise ValueError('can not interpret formula when design'
```
Should have been checked somewhere else

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,451 @@
+                                 ' ,"p_value", "effect_size" or "effect_variance"')
+        else:
+            raise ValueError('output_type must be one of "z_score", "stat",'
+                             ' "p_value","effect_size" or "effect_variance"')
```
missing space after comma

bthirion commented on nistats/tests/test_first_level_model.py:
```diff
@@ -296,6 +296,16 @@ def test_first_level_model_contrast_computation():
+        model.compute_contrast('c2')
         # smoke test for one null contrast in group
         model.compute_contrast([c2, cnull])
         # only passing null contrasts should give back a value error
```
Could be in another PR ?

bthirion commented on nistats/tests/test_second_level_model.py:
```diff
@@ -0,0 +1,244 @@
+        assert_raises(ValueError, model.compute_contrast, [])
+        # fit model
+        Y = [func_img]*4
+        X = pd.DataFrame([[1]]*4)
```
pep8...

bthirion commented:
That's all for now. We can discuss on design aspect. I think we will be mreo efficient if we can remove from this PR the aspects that are not related to second-level model: first-level model improvement and version things.
Thanks !

MartinPerez commented:
I fixed all comments, created an example and improved documentation of the second level design matrix and eliminated first_level_conditions_name by making first_level_conditions a list of tuples. I do not use a dictionary to be able to provide the conditions in a given order to match to design matrix rows, which is a proposed user case.

This depends on the dependencies and first_level_upgrades branches. When those are merged, I will rebase and update this PR.

MartinPerez commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,381 @@
+                    if condition not in maps_table['map_name'].tolist():
+                        raise ValueError('condition %s not present in'
+                                         ' second_level_input' % condition)
+                condition_list = list(zip(*first_level_conditions))[:][1]
```
this is the ugly way of handling zip in python 3 :S. Any advice?

bthirion replied:
No.

MartinPerez replied:
tlooking again at this, would not be more readable:?

[condition[0] for condition in first_level_conditions]

no idea what passed my mind when I used zip here

MartinPerez commented:
@bthirion rebased done, all comments addressed and tests passed.

bthirion commented on examples/plot_mixed_gambles.py:
```diff
@@ -0,0 +1,81 @@
+
+    # Create snapshots of the contrasts
+    display = plotting.plot_stat_map(z_map, display_mode='z',
+                                     threshold=3.0, title=contrast_id)
```
These maps are desperately empty. Isn't there a contrast that displays a significant effect ?

bthirion commented on examples/plot_mixed_gambles.py:
```diff
@@ -0,0 +1,81 @@
+df_columns = [subjects_id, behavioral_target, zmap_paths]
+df_columns = zip(*df_columns)
+df_column_names = ['model_id', 'map_name', 'effects_map_path']
+first_level_df = pd.DataFrame(df_columns, columns=df_column_names)
```
It seems clearer to me to call it `second_level_df`.

bthirion commented on examples/plot_mixed_gambles.py:
```diff
@@ -0,0 +1,81 @@
+# create input for second-level analysis
+df_columns = [subjects_id, behavioral_target, zmap_paths]
+df_columns = zip(*df_columns)
+df_column_names = ['model_id', 'map_name', 'effects_map_path']
```
'map_name' sounds a bit strange: isn't it simply the experimental condition ?

bthirion commented on examples/plot_second_level_design_matrix.py:
```diff
@@ -0,0 +1,46 @@
+maps_model = subject_list * 2
+maps_path = [''] * n_subjects * 2
+maps_table = pd.DataFrame({'map_name': maps_name,
+                           'model_id': maps_model,
```
I don't like `model_id`. Maybe rather `subject_id` ?

bthirion commented on nistats/design_matrix.py:
```diff
@@ -446,3 +446,47 @@ def plot_design_matrix(design_matrix, rescale=True, ax=None):
+
+
+def create_second_level_design(maps_table, regressors=None):
+    """Infers a second level design from a maps table.
```
Sets up a second level design...

bthirion commented on nistats/design_matrix.py:
```diff
@@ -446,3 +446,47 @@ def plot_design_matrix(design_matrix, rescale=True, ax=None):
+        The second level design matrix
+    """
+    maps_name = maps_table['map_name'].tolist()
+    subjects_id = maps_table['model_id'].tolist()
```
I think that we agree on renaming model_id to subject_id...

bthirion commented on nistats/first_level_model.py:
```diff
@@ -311,17 +312,17 @@ def fit(self, run_imgs, events=None, confounds=None,
         design_matrices: pandas DataFrame or list of pandas DataFrames,
             Design matrices that will be used to fit the GLM. If given it
-            takes precedence over events and confounds.
+            takes precedence over events and regressors.
```
Why do you prefer `regressors` over `confounds` ? What I like with confounds is that it makes a clear distinction wrt the other regressors of interest.

salma1601 replied:
Does this `confounds` plays the role of `add_regs` in the nipy design matrix ?

bthirion replied:
Yes.

salma1601 replied:
Does it really need to be a confound ? For instance for ASL we can include additionnal regressors by convolving the PRF with the conditions. May be the terminology covariate can be used here ?

bthirion replied:
The best answer is that we simply don't cover ASL for the moment. If you would like to contribute it, this should come as a specific PR. I think that it's necessary to have modality specific models if we want to obtain a user-friendly API.

salma1601 replied:
> If you would like to contribute it, this should come as a specific PR.

Ok, may be I can give that a try after some time.

>Note however that in spite of the 'confounds' term, these regressors are actually handled as well as the others in the glm and that you can make inference upon them if you wish or need.

Yes, that was the sens of my question. I found the word `confounds` a little misleading for it may suggest the opposite of what you are pointing to (actually it frightened me w.r.t. some old analyses, although I was almost sure all the regressors are treated the same way). To be honest I personally preferred the word `regressors`, or at least `covariates`. On the other hand I agree with modality specific models strategy.

MartinPerez commented:
Reverted model id to subject id and regressors to confounds. Changed example of mixed gambles to button press contrasts from localizer. I would add a BIDS example to show firstlevel-secondlevel more complex setup in the next BIDS utils PR.

MartinPerez commented:
the only point not addressed yet is the 'map_name', 'map_effect_size', 'map_effect_variance' names in the maps table. I think "map" is the most general term that is not misleading. Would you suggest a different word? Otherwise I handled what we discussed and latest comments.

bthirion commented on examples/plot_second_level_button_press.py:
```diff
@@ -0,0 +1,70 @@
+
+1. A sequence of subject fMRI button press contrasts is downloaded.
+2. a mask of the useful brain volume is computed
+3. A GLM is applied to the dataset (as fixed effects, then contrast estimation)
```
A one-sample t-test is applied to the brain maps.

bthirion commented on examples/plot_second_level_button_press.py:
```diff
@@ -0,0 +1,70 @@
+                             columns=['left-right'])
+
+# Display subject t_maps
+subjects = [s[0] for s in data['ext_vars']]
```
subject_data[0] for subject_data in data['ext_vars']

bthirion commented on examples/plot_second_level_button_press.py:
```diff
@@ -0,0 +1,70 @@
+                                    plot_abs=False, display_mode='z')
+display.savefig(os.path.join(write_dir, 'left-right_group.png'))
+
+plotting.show()
```
This example is great, thanks.
Maybe another one that tests a differential effect (e.g. handedness) would help too.

MartinPerez replied:
I would prefer to focus on sphinx, circle ci before developing more the tests. I will post today the barebones sphinx generated webpage.

bthirion replied:
OK, I'm going to open an issue for that.

bthirion commented on examples/plot_second_level_design_matrix.py:
```diff
@@ -0,0 +1,46 @@
+                           'effects_map_path': maps_path})
+# Specify extra information about the subjects to create additional regressors
+extra_info_subjects = pd.DataFrame({'subject_id': subject_list,
+                                    'age': [i for i in range(15, 35)],
```
`'age': range(15, 35),`
should work too.

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,381 @@
+        Mask to be used on data. If an instance of masker is passed,
+        then its mask will be used. If no mask is given,
+        it will be computed automatically by a MultiNiftiMasker with default
+        parameters. Mask automatic computation assumes first level imgs have
```
Automatic mask computation

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,381 @@
+        fit. By default, no caching is done. Creates instance of joblib.Memory.
+
+    memory_level: integer, optional
+        Rough estimator of the amount of memory used by caching. Higher value
```
This docstring is a bit hard to follow. When I read "Rough estimator of the amount of memory used by caching" I start thinking "My data are bout 300MB", which is not what is expected. The expected value is typically 1.

MartinPerez replied:
This is related to the masker and I am not sure myself of what the memory level is regulating. I agree the documentation is not helpful. This also appears in other parts of the code.

MartinPerez replied:
I opened an issue for this.

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,381 @@
+    n_jobs : integer, optional
+        The number of CPUs to use to do the computation. -1 means
+        'all CPUs', -2 'all CPUs but one', and so on.
+
```
'minimize_memory' should be documented.

MartinPerez replied:
done

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,381 @@
+        second_level_input: list of `FirstLevelModel` objects or pandas
+                            DataFrame or list of Niimg-like objects.
+            If list of `FirstLevelModel` objects, then first_level_conditions
+            must be provided. If a pandas DataFrame, then must contain columns
```
...then they have to contain...

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,381 @@
+
+            If second_level_input is a list of Niimg-like objects then this
+            argument is ignored.
+
```
This docstring is also hard to read, because it describes a complex logic. This is where a series of examples for model specification will be useful

MartinPerez replied:
I agree. Just opened an issue for this.

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,381 @@
+                if first_level_conditions is None:
+                    raise ValueError('First level models input requires'
+                                     'first_level_conditions to be provided')
+                for midx, first_level_model in enumerate(second_level_input):
```
`midx` -> 'model_idx' would be more readable.

bthirion commented on nistats/second_level_model.py:
```diff
@@ -0,0 +1,381 @@
+
+        # check design matrix X and effect maps Y agree on number of rows
+        if len(effects_maps) != design_matrix.shape[0]:
+            raise ValueError('design_matrix X number of rows do not agree with'
```
does not match the number of maps considered

bthirion commented on nistats/tests/test_second_level_model.py:
```diff
@@ -0,0 +1,222 @@
+        model.compute_contrast(c1, 'F', 'p_value')
+        model.compute_contrast(c1, 'F', 'effect_size')
+        model.compute_contrast(c1, 'F', 'effect_variance')
+        # formula should work (passing varible name directly)
```
variable

bthirion commented:
All right, the PR looks good. With only tiny details remaining, it should be mergeable soon.

MartinPerez commented:
I addressed comments. I am posting today the sphinx circle ci branch to start handling previous and new examples well.

bthirion commented:
I'll do another review today or tomorrow. We're close !

MartinPerez commented on nistats/design_matrix.py:
```diff
@@ -446,3 +446,47 @@ def plot_design_matrix(design_matrix, rescale=True, ax=None):
+    if len(np.unique(design_columns)) != len(design_columns):
+        raise ValueError('Design matrix columns do not have unique names')
+
+    return design_matrix
```
so should I add a check like this here?:

linalg.cond(x) < 1/sys.float_info.epsilon

bthirion replied:
yes. At this point I would probably spit a warning rather than an Error -- it is clear that using such a design matrix will lead to aberrant estimates.

MartinPerez commented:
I made the changes, need to kickstart travis.

MartinPerez commented:
my bad, travis is fine. All tests passed. circle ci is of course meant to fail in this branch.

bthirion commented:
All right, merging
